tests/iter.rs

```diff
@@ -7,6 +7,27 @@ use std::ops::Drop;
 use generic_array::typenum::consts::U5;
 use generic_array::GenericArray;
 
+#[test]
+#[should_panic(expected = "ExactSizeIterator lied about its length")]
+fn test_from_iterator() {
+    struct BadExact(usize);
+
+    impl Iterator for BadExact {
+        type Item = usize;
+        fn next(&mut self) -> Option<usize> {
+            if self.0 == 1 {
+                return None;
+            }
+            self.0 -= 1;
+            Some(self.0)
+        }
+    }
+    impl ExactSizeIterator for BadExact {
+        fn len(&self) -> usize { self.0 }
+    }
+    let _array = GenericArray::<usize, U5>::from_exact_iter(BadExact(5)).unwrap();
```

Comment:
Without this PR the test would not panic, but leave `_array` with its last element uninitialized.